Changes to add density following adjoint feature

ciarapimm opened the pull request:
## What changes does this PR introduce?
This PR adds the changes needed to introduce a density following adjoint feature.

## What is the current behaviour? 
#483 - this issue explains how the new feature is intended to work. 

## Does this PR introduce a breaking change? 
No

## Other information:



## Suggested addition to `tag-index`
(To avoid unnecessary merge conflicts, please don't update `tag-index`. One of the admins will do that when merging your pull request.)

mjlosch commented:
@ciarapimm 
I tried this with your https://github.com/ciarapimm/verification_other/tree/cp_testcase/global_oce_cs32 and there are extensive recomputation warnings associated with: `The responsible variables are : ac2d,as2d,aw2d,pc,ps,pw `
and `main_do_loop` is called again within a second loop over `ilev_1` (which is the innermost checkout loop counter). Also (probably follow on) `do_oceanic_phys` is called again from `forward_step_ad`, which we would like to avoid. Do you see that too?

ciarapimm commented:
Thank you @mjlosch for your very helpful comment. 

Could you please clarify to me in a bit more detail where you find these warnings, as I am not sure when this is happening?

mjlosch commented:
After `make adtaf` (part of `make adall`) to generate the AD-code, you'll get a `taf_ad.log` file that contains a lot of information about the code generation. Particularly interesting are these RECOMPUTATION warnings. Some of the are "benign" and we tend to ignore them, but some are (very) expensive, especially when they appear in `the_main_loop.f`. You will also see in the AD-code, that for each inner loop round there are `ilev_1` additional calls of `main_do_loop` (basically the entire model time step). So you call the model (N!) times too often (if N=nchkptlev_1) in each of the nhkptlev_2*nckptlev_3 loop rounds.

Compare the structure of the_main_loop_ad and forward_step_ad (and the taf_ad.log files) between upstream/master and this PR, you'll see the difference immediately. "testreport" now counts, how often do_oceanic_phys (dop) and load_fields_driver (led) are called. Both should be 1 only. Again, compare between upstream/master and this PR.

mjlosch commented:
For the record, this is what I did to download and compile this branch and the test experiment:
```
git remote add ciarapimm https://github.com/ciarapimm/MITgcm.git
git fetch ciarapimm
git checkout --track ciarapimm/cp_issue483
git clone -b cp_testcase git@github.com:ciarapimm/verification_other.git

cd verification_other
../verification/testreport -adm -ncad -devel -t global_oce_cs32
```
The run will fail because I did not download any forcing files etc., but I am only interested in the compilation at this point. This leads to a `taf_ad.log` that contains at least 9 new recomputations warnings, all in `the_main_loop`, here's the first one:
```
TAF RECOMPUTATION LOOP WARNING DOLOOP_STMT the_main_loop.for:3114 in the_main_loop
extensive recomputations are required.
The responsible variables are : ac2d,as2d,aw2d,pc,ps,pw
```
I think one way forward is to find places where the fields `ac2d,as2d,aw2d,pc,ps,pw` are already stored and adjust the corresponding cpp-flags. 

The next question is why these fields need storing after your changes: They are the matrix coefficients and preconditioner coefficients of the 2D barotropic pressure solver (cg2d.F) and are already stored with NONLIN_FRSURF defined (which is the case in the global_oce_cs-experiment). It's not clear why your relatively simple changes have such a large effect on the recomputation/flow analysis. Maybe you can try to precompute a 2D density field per player before computing the mask?

mjlosch commented:
@ciarapimm after unsuccessfully fiddling with your code for quite some time, I took a leap (in German "Flucht nach vorn"= something like "running away towards the enemy lines") and merged the upstream/master into this branch. This fixed all recomputation problem that I had. I don't want to push this merge to your PR, because I had some issue during the merge related to ecco.h, which has changed to upper case ECCO.h, a difference that my MacOS computer does not appreciate. So could you please try and merge the upstream master into your branch and push it? Instructions are at the end of the [quickstart guide section in the manual](https://mitgcm.readthedocs.io/en/latest/contributing/contributing.html#quickstart-guide). During the merge, you'll have to carry over your changes to ecco.h into ECCO.h, otherwise they will be lost.

ciarapimm commented:
Thanks @mjlosch! When we tested this ourselves using https://github.com/ciarapimm/verification_other/tree/cp_testcase/global_oce_cs32 , I would run two short tests one with gencost_useDensityMask(1) = .TRUE., and one with gencost_useDensityMask(1) = .FALSE.,. in data.ecco. If you set the density range to be very large in data.ecco, then the cases with the density following feature on and off will give approximately the same results. This is so the density following feature makes sure to keep all of the depths you want and will effectively be the same as running without it. I hope this all makes sense, I'm not sure I have explained myself very well. Please let me know if you have any questions?

owang01 commented:
@ciarapimm Thank you for the PR. I think it is a useful feature. 

I corrected a couple of minor bugs in ECCO.h of the PR that caused compilation errors. The new gencost arrays should be of size NGENCOST and defined outside of the CPP option ALLOW_GENCOST3D. See my updated ECCO.h https://github.com/owang01/MITgcm/blob/2c7a5f8d6c92f97d8d4e15071dcf5807d29acf25/pkg/ecco/ECCO.h.

I was not able to run your test case https://github.com/ciarapimm/verification_other/tree/cp_testcase/global_oce_cs32. The model stopped when calling ecco_div in ecco_physics.F because areavolGlob is zero (areavolGlob is used as the denominator in the ecco_div subroutine). The reason areavolGlob is zero is probably because the size of the masking file density_mask_SAUSC https://github.com/ciarapimm/verification_other/blob/6305b185c16027b7689854bfb2435aaf675d683a/global_oce_cs32/input_ad.sens/density_mask_SAUSC (421200 bytes) is not consistent with SIZE.h in the code directory https://github.com/ciarapimm/verification_other/blob/6305b185c16027b7689854bfb2435aaf675d683a/global_oce_cs32/code/SIZE.h#L45-L55.

mjlosch commented on pkg/ecco/ecco_phys.F:
```diff
@@ -253,38 +246,64 @@ subroutine ecco_phys( myThid )
+            ENDIF
+
+c
+c ---- If density mask is enabled, use it here ----
```
@ciarapimm I have a few questions/comments about the coding style here. 
- Does is make sense to use salinity as a mask? Even if it is unlikely, it is possible to have zero salinity in the ocean and the equation of state is still defined. I think it would be better to use hFacC or maskC, e.g. hFacC .GT. 0.0
- not that important: one could contract the two if statements into `IF ( gencost_useDensityMask(kgen) .AND. hFacC(i,j,k,bi,bj).GT.0. _d 0) THEN`...
- we usually do not mix floats and integers, so instead of l294 you should write `tmpsig = tmpsig - 1000. _d 0`, 
- l297 and 299: multiplying by 0.5 is more efficient than diving by 2.0

mjlosch replied:
Just a suggestion: one could also simplify the expression a little, e.g. 
```
              tmpsig_lower = 0.5 + 0.5*tanh(gencost_tanhScale(kgen)
     &           *(tmpsig-gencost_sigmaLow(kgen)))
              tmpsig_upper = 0.5 - 0.5*tanh(gencost_tanhScale(kgen)
     &           *(tmpsig-gencost_sigmaHigh(kgen)))
```

ciarapimm replied:
Hi @mjlosch. 

Thank you for these comments on the style of coding. I have now updated the ecco_phys.F file with these changes.

mjlosch commented:
@ciarapimm I also tried to run the test case an get the same error as Ou. I think it is related to the conflict in pkg/ecco/ecco_phy.F: The argument list has changed, and somehow this has not been carried over in the merge, probably a consequence of my "cleaning up that file". Anyhow, after merging again with upstream/master, the problem goes away, the input_ad.sens finishes but with zero cost and hence zero sensitivities. Can you reproduce that?

@owang01 As far as I understand, only density_mask_SAUST is used in the test, but I don't understand these generic cost function things very well.

mjlosch commented:
Sorry, I have to correct myself: after merging with latest code ecco_div actually checks numerLoc and if it is zero, the routine just returns the input field. So @owang01 is probably correct about the masks. Let's not merge upstream/master into this branch before these things are sorted out.

ciarapimm commented:
Hi @owang01 and @mjlosch ,

Thank you both so much for all of your help with this. Thank you for your comments, and the useful edits to the ECCO.h and ecco_phys.F files. These files have now bee updated following on from both of your suggestions.  

Looking at the comments on running the test case, I did find the same problem with the mask (density_mask_SAUSC) that I used in /verifcation_other/cp_testcase/. This was my mistake as it was a copy of a mask I used in a ECCOv4r2 test, so as they have different grid sizes, I think that was the problem. 

However, I have now tested using the south30_mask (already included in the global_oce_cs32 case) which is in the correct grid format and I can't find any problems with using this. So when using the correct size mask I believe this feature should be working again.

mjlosch commented:
@ciarapimm thanks for responding to my suggestions. Have you updated the verification_other/cp_testcase so that it uses the correct mask, yet? If not, please do so, so that we/I can also test this experiment.

Also, it may be a good idea to merge the current upstream/master and resolve the conflict (
```
git checkout master
git pull upstream master
git checkout cp_issue483
git merge master
```
 ... and then resolve conflicts).

I haven't found a solution to the other problem (dop=2) yet. I don't see why these extra lines of code require extra information, so that the forward_step is being recomputed again. Do you? Unless we find a fix for that, there should be a CPP flag "ifdef ECCO_ALLOW_DENSITY_MASK" or similar that is undefined by default to avoid the extra recomputations when they are not needed.

ciarapimm commented:
@mjlosch 

Yes, I have updated data.ecco in /verification_other/cp_testcase/ to use the mask south30_mask. So you should hopefully be able to run this now.

I think I have now also resoled the conflict in ecco_phys.F, I kept the master branch. 

For the dop=2 issue, I'm not sure where this comes from. Could you let me know where you found this please?

mjlosch commented:
@ciarapimm 
after running `make adtaf` in the build directory you can count how often "do_oceanic_phys" is called. If it is called more often than once, it usually means that it is called from "forward_step_ad", which shouldn't happen. I use the testreport flag "-ncad", i.e. the genmake2-flag "-nocat4ad" to get separate files for each input, and then it is straightforward to inspect "forward_step_ad.f": `grep do_oceanic_phy forward_step_ad.f` should give you:
```
      call do_oceanic_phys_ad( mytime,myiter,mythid )
      call do_oceanic_physmd( mytime,myiter,mythid )
      call do_oceanic_phys( mytime,myiter,mythid )
```
but with this PR is gives:
```
      call do_oceanic_phys( mytime,myiter,mythid )
      call do_oceanic_phys_ad( mytime,myiter,mythid )
      call do_oceanic_physmd( mytime,myiter,mythid )
      call do_oceanic_phys( mytime,myiter,mythid )
```
You'll see that there are also extra calls to thermodyamics, solve_for_pressure, etc, that I got rid off with painfully inserting store directives at the right places. With your changes, they are all back. It probably has to do with the sea ice model, but I don't know why, yet.

mjlosch commented:
The AD problem is fixed with e702861 and we are back to dop=1.
The code looks good now, maybe @owang01 can have another look at this.

@ciarapimm can you add a little bit of documentation of the new parameters, e.g. [here:](https://mitgcm.readthedocs.io/en/latest/ocean_state_est/ocean_state_est.html#ecco-model-data-comparisons-using-gridded-data-sets), and a suggestion for the tag-index (top of this page)?
Also, will there be an example experiment? We could have this added to verification/ab_sea or verification_other/global_oce_cs32 as an additional cost function contribution, or replacing one of the cost functions). This would change results, but I wouldn't mind.

ciarapimm commented:
@mjlosch Thank you for the edits fixing some of the issues. 

How would I add the edits to the documentation, please? I think I know what I would want to add. 

For the example experiment I think it would be good to have it added to verification_other/global_oce_cs32/, as this is where I have done most of the testing.

mjlosch commented:
For the documentation, you could have a look at [chapter 10](https://mitgcm.readthedocs.io/en/latest/ocean_state_est/ocean_state_est.html) of the online documentation, in particular the section about the generic cost function (10.1.1, I believe) and maybe add your parameters to a table or similar. @jrscott will most likely be able to help you with this. The file to edit for that is doc/ocean_state_est/ocean_state_est.rst. In [Section 5.6](https://mitgcm.readthedocs.io/en/latest/contributing/contributing.html#contributing-to-the-manual) provides more instructions.

I would be great, if we could add this functionality to lab_sea, but @owang01 may have a look at this and may come back to you.

ciarapimm commented:
I have now added my edits to the documentation. I added the new parameters to the corresponding table, and have then added a new paragraph explaining how to use the density following feature. 

I have also added some suggestions to the tag-index at the top of this page.

mjlosch commented:
@ciarapimm thanks, the documentation looks good to me, I add small fixes, please check. 
Since I don't understand this masking business, it's not clear to me how it works in this case either.
Last thing to to: add this to a verification experiment. It's not clear to me how simple it would be to add this to lab_sea (or how useful), but in the meantime, maybe @ciarapimm could cleanup  https://github.com/ciarapimm/verification_other/tree/cp_testcase/global_oce_cs32 (comment back in gencost 2 and 3, make sigmaLow/High really "reals", i.e. "23." instead of "23", as some compilers still complain about that, and clean up the masking properly and remove unnecessary files) and create a separate PR for this. What do you think?
Maybe it's a good idea to have a separate 4th gencost contribution for this new feature (which is a copy of the first, but with the density parameters added)? But that would need to be discussed in the separate PR.

ciarapimm commented:
@mjlosch Thank you for the changes to documentation, they look good to me. 

I had a look at how to add this to verification/lab_sea but as I have not used it before I found it quite hard to do. I don't really understand how the divided adjoint works properly. In theory I don't think it would be too difficult to add but I'm just not entirely sure how to do it. 

I have now tried to clean up cp_testcase, and opened a new draft PR: Example experiment using the density following adjoint feature #51.

jrscott commented:
doc change LGTM

owang01 reviewed (changes requested):
@ciarapimm It looks good to me. 

Per @mjlosch's suggestion, I added this feature to the lab_sea experiment. Since pushing the changes to your PR was denied, I put the changes in https://github.com/owang01/MITgcm/tree/cp_issue483. 

The changes include: 1) modifying lab_sea/input_ad/data.ecco to add a new cost using the density following feature, 2) adding two binary files also in lab_sea/input_ad/ for horizontal masking (2d; labsea_densityfollowing_maskC) and vertical masking (1d; labsea_densityfollowing_maskK). 

I set mult_gencost = 1.D+02 for the new cost so it would have similar magnitude to the other existing large costs in the two adjoint experiments "ad" and "noseaicedyn". In the adjoint experiment "noseaice", however, setting mult_gencost = 1.D+02 would make the new cost two orders larger than the existing costs. One alternative of course is to create another data.ecco in input_ad.noseaice with a smaller mult_gencost (=1) for this new cost.

owang01 commented:
@jm-c Thanks for the suggestions. I'll create a new PR with the example using the density following adjoint feature.

ciarapimm commented:
@mjlosch Is there now anything else that this PR requires?

ciarapimm commented:
Hi @jm-c , Thank you for your comments! 
Just a note on the first point, I thought we were hoping to have the new feature in both experiments, "lab_sea" and "global_oce_cs32" ?
Excited to get this feature merged though!

jm-c commented:
@mjlosh Since commit 5659757a23479073ff1554c1ce4060ac99f0bd99 intorduces many (minor) differences in ecco_phys.F, I would like to finish the cleaning and update the S/R description in the standard way, so that we don't have to come back to change it again later. What do you think of this ?

mjlosch commented:
@jm-c I am fine with that.
